[MU4] Check for corruption before attempting to save

Eism opened the pull request:
Resolves: #16042

cbjeukendrup commented on src/engraving/libmscore/undo.cpp:
```diff
@@ -292,11 +292,7 @@ UndoStack::UndoStack()
-    for (auto c : list) {
-        c->cleanup(idx++ < curIdx);
-    }
-    DeleteAll(list);
```
Where has `DeleteAll(list);` gone in `UndoStack::clear()`? It looks like we would still need it to prevent memory leaks, don't we?

cbjeukendrup commented on src/engraving/libmscore/cmd.cpp:
```diff
@@ -331,6 +331,24 @@ void Score::undoRedo(bool undo, EditData* ed)
+
+    auto undoStack = this->undoStack();
+    while (!undoStack->isClean()) {
+        undoStack->undo(ed);
```
This seems to me a very risky way of reverting to the last saved state. If the score has become corrupted, I would trust the UndoStack even less. In some cases with corruptions or crashes that I've worked on, it looked indeed like undoing only made it worse. 
I thought we would just close and completely re-read the file from disk, to remove any risk.

Eism replied:
yep, fixed

RomanPudashkin commented on src/project/internal/projectactionscontroller.cpp:
```diff
@@ -378,58 +379,43 @@ IInteractive::Button ProjectActionsController::askAboutSavingScore(INotationProj
     if (!project->isNewlyCreated()) {
         if (project->isCloudProject()) {
-            return saveProjectToCloud(project->cloudInfo());
+            return saveProjectAt(SaveLocation(SaveLocationType::Cloud, project->cloudInfo())); // todo force
```
Is this TODO still needed?

RomanPudashkin commented on src/project/internal/projectactionscontroller.cpp:
```diff
@@ -938,15 +918,130 @@ void ProjectActionsController::warnPublishIsNotAvailable()
+
+void ProjectActionsController::warnScoreCorruptionUponOpenning(const std::string& errorText)
+{
+    std::string title = trc("project", "Your score cannot be saved");
```
warnScore**CorruptionUponOpenning** but "Your score cannot be **saved**"

Wrong text?

cbjeukendrup commented on src/notation/view/internal/undoredomodel.cpp:
```diff
@@ -67,6 +67,10 @@ void UndoRedoModel::load()
     });
 
+    context()->currentProjectChanged().onNotify(this, [this]() {
+        load();
```
I think we can simply call `emit stackChanged();` here. No need to re-subscribe to notifications, if I'm not mistaken.

cbjeukendrup commented on src/project/internal/projectactionscontroller.cpp:
```diff
@@ -938,15 +918,145 @@ void ProjectActionsController::warnPublishIsNotAvailable()
+    }
+}
+
+void ProjectActionsController::warnCorruptedScoreUponOpenningCannotBeSaved(const std::string& errorText)
```
I'm not sure whether this method is following the design...
When the score _was already corrupted_ at the time that it was opened:
- saving locally should just work without any warning;
- saving/publishing to the cloud should be prohibited.

It looks like this method also prevents saving the score locally; and the title should be `Your score cannot be uploaded to the cloud` (and the "Cancel" button should be an "OK" button).

Excuse me if this is not true or if the design has been changed without that I know.

cbjeukendrup commented on src/project/internal/projectactionscontroller.cpp:
```diff
@@ -938,15 +918,145 @@ void ProjectActionsController::warnPublishIsNotAvailable()
+
+    int defaultBtn = saveCopyBtn.btn;
+
+    IInteractive::ButtonData revertToLastSavedBtn(saveCopyBtn.btn + 1, trc("project", "Revert to last saved"),
+                                                  true /*accent*/);
```
Aren't we forgetting `buttons.push_back(revertToLastSavedBtn);`?

cbjeukendrup commented on src/project/inotationproject.h:
```diff
@@ -58,8 +58,10 @@ class INotationProject
 
     virtual void markAsUnsaved() = 0;
 
+    virtual void revertChanges() = 0;
```
It looks like we're not using this method because it's fully handled in `ProjectActionsController::revertCorruptedScoreToLastSaved`.

cbjeukendrup commented on src/project/internal/projectactionscontroller.cpp:
```diff
@@ -938,15 +918,145 @@ void ProjectActionsController::warnPublishIsNotAvailable()
+    }
+}
+
+void ProjectActionsController::revertCorruptedScoreToLastSaved()
```
I'm missing this dialog from the design: 
<img width="499" alt="Scherm­afbeelding 2023-02-01 om 13 24 58" src="https://user-images.githubusercontent.com/48658420/216041861-399dc1a0-f04f-445b-9379-abd09e686fb2.png">

cbjeukendrup commented:
For testing: a convenient deterministic way of corrupting a score is #16151. (Since #15935, the score is indeed detected as corrupted in that case.)

(@RomanPudashkin so for that purpose, it would be nice to wait before merging #16165 until this pull request is fully tested 🙂)

Eism commented:
> (@RomanPudashkin so for that purpose, it would be nice to wait before merging https://github.com/musescore/MuseScore/pull/16165 until this pull request is fully tested 🙂)

Yes, I used this case for testing. 
No need to block the merging of 16165, I will not rebase the master until the end of testing

bkunda commented:
@Eism tested on Win and macOS. Just a few things: 

1. When saving locally after a save-to-cloud attempt (using the ‘save a copy‘ button): MuseScore doesn't seem to recognise that the score has been saved, even though a file has actually been created on the computer [(See demo video)](https://www.dropbox.com/s/ty9xix9v8i88pf5/save-copy-issue.mov?dl=0).
2. When attempting to save/publish to cloud an _already corrupted score_, this dialog is inconsistent with the design: 

_Current implementation_
<img width="526" alt="Screenshot 2023-02-06 at 12 04 49 pm" src="https://user-images.githubusercontent.com/86290556/216959214-31b6abbf-22eb-444b-82e3-0f9cc9d8a87a.png">

_Design_
<img width="555" alt="214262751-22db8ee7-91bb-4dfa-aff1-5390871a4e05" src="https://user-images.githubusercontent.com/86290556/216959393-3d5b13b4-7379-499d-9e35-24d861ce9996.png">

3. A minor typo in this dialog (actually an error in my design):
<img width="426" alt="Screenshot 2023-02-06 at 12 08 23 pm" src="https://user-images.githubusercontent.com/86290556/216959716-b45ccf59-bc51-4b25-91bc-e7bf6e4dec8f.png">

The last sentence should read **This action cannot be undone** (I.e. "cannot" instead of "can not"). 

Thank you!

Eism commented:
@bkunda thanks!

1. the problem is also in master. @cbjeukendrup, your [commit](https://github.com/musescore/MuseScore/commit/de3457#diff-ba955a69c6d48fc8e1ad2625cdaf6abeffa16d1fd53774ebe5c57d7aabe069dfL370) is the cause of this bug. I'll leave it for you. @bkunda please create new issue
2. fixed. But the Ok button is an accent
3. fixed

cbjeukendrup commented:
Also, I think the "Save a copy" (or "Save as") button in the corruption dialog should immediately open the "save locally" dialog, because it does not make sense to suggest saving to the cloud if we already know that the score is corrupted.

bkunda commented:
> What you're expecting rather sounds like "Save as":

Yes on reflection I think this might actually make more sense. The behaviour of "Save a copy" will probably stump most users (it obviously stumped me!), and their main goal if they go for this option is most likely going to be to send their corrupted file to someone on the forum for assistance, so the behaviour should be as familiar as possible. 

So I guess this should probably be something like "Save to computer". If the file was already saved, it would perform a regular save function. If the file was not already saved, it would trigger the local "Save as" dialog. 

The UI copy would then be updated to something like: "You can also save the score locally and get help to fix the errors on musescore.org". 

> Also, I think the "Save a copy" (or "Save as") button in the corruption dialog should immediately open the "save locally" dialog, because it does not make sense to suggest saving to the cloud if we already know that the score is corrupted.

Completely agree with this 🙂

bkunda commented:
Hi @Eism,

So one change to the specification please: 

When the user attempts to save a "clean" (i.e. uncorrupted) score to the cloud _which has become corrupted since it was opened_, they see this dialog: 
 
<img width="596" alt="Group 14192" src="https://user-images.githubusercontent.com/86290556/218058689-24212701-c505-493a-9b8f-3a899e1e1122.png">

The **Save As...** button would immediately trigger the regular operating system "save" dialog, _not_ MuseScore's "how would you like to save?" dialog.

Hope this is clear 🙂 Let me know when it's ready for further testing.

Eism commented:
@bkunda 
Fixed. Please test this build https://github.com/Eism/MuseScore/actions/runs/4143485042

(Builds in this PR don't start due to conflicts)

bkunda commented:
@Eism thanks! Just one more thing I've found (actually another gap in the design – my apologies).

For a new score that has _never been saved before_, and which becomes corrupted while the user is working: when the user attempts to save the score to the cloud for the first time, they currently see this dialog (in your build):

<img width="528" alt="Screenshot 2023-02-10 at 4 24 56 pm" src="https://user-images.githubusercontent.com/86290556/218131327-9e5c8ba4-0c2c-42e5-bf20-1271cb012dff.png">

The options are right, but the messaging doesn't make sense (my oversight here). Kindly please change the body copy to: 

"This score has become corrupted and contains errors. You can fix the errors manually, or save the score to your computer and get help for this issue on musescore.org."

Thank you!

Eism commented:
@bkunda 
Fixed, but also rebased to the current master, so I'm not sure if you can make a score corrupted.

bkunda commented:
> @bkunda
> Fixed, but also rebased to the current master, so I'm not sure if you can make a score corrupted.

Yeah I just noticed this. No worries, I shall take your word that this last fix has been made 😀. Many thanks!
